Ncgenerics test fixes kavon v10

kavon opened the pull request:
more test fixes

kavon committed:
ConformanceLookup: filter ReferenceStorageType

There's a few uses of ReferenceStorageTypes being substituted for
generic parameters, at least in the test suite, such as
`Optional<@sil_unmanaged ..>` and just plain `<@sil_unowned ..>`.

Before, conformance lookup could (and did) give bogus answers when asked
 if the type satisfies any conformance requirements. Now with
 NoncopyableGenerics, we will interpret such conformance lookups as
 being asked of the referent type, ignoring the SIL ownership wrapping
 it.

kavon commented:
preset=ncgenerics,smoketest=macosx
@swift-ci Please test with preset macOS Platform

kavon committed:
NFC: restore an assert I accidentally removed

kavon commented:
@swift-ci smoke test

kavon commented:
preset=ncgenerics,smoketest=macosx
@swift-ci Please test with preset macOS Platform

slavapestov commented on lib/AST/ConformanceLookup.cpp:
```diff
@@ -586,6 +586,20 @@ LookupConformanceInModuleRequest::evaluate(
+    if (auto ip = protocol->getInvertibleProtocolKind()) {
+      ImplicitKnownProtocolConformanceRequest req{nominal,
+                                                  getKnownProtocolKind(*ip)};
+      if (auto conformance = evaluateOrDefault(ctx.evaluator, req, nullptr))
```
Can the request just return a ProtocolConformanceRef

slavapestov commented on lib/AST/GenericEnvironment.cpp:
```diff
@@ -417,6 +417,9 @@ Type GenericEnvironment::mapTypeIntoContext(GenericEnvironment *env,
 }
 
 Type MapTypeOutOfContext::operator()(SubstitutableType *type) const {
+  if (auto genericParam = dyn_cast<GenericTypeParamType>(type))
```
Bug

slavapestov commented on lib/Sema/TypeCheckInvertible.cpp:
```diff
@@ -349,16 +349,26 @@ ProtocolConformance *deriveConformanceForInvertible(Evaluator &evaluator,
+      // superclasses, such as:
+      //    class Base<T>: IP {}
+      //    class Derived: Base<Derived> {}
+      return ctx.getBuiltinConformance(nominal->getDeclaredInterfaceType(),
```
the conformance should be formed in lookupConformance() with the conforming type, not the nominal's declared interface type. Generally we want this to hold:

```
lookupConformance(type, proto)->getType()->isEqual(type)
```

slavapestov commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -4864,6 +4864,11 @@ static void ensureRequirementsAreSatisfied(ASTContext &ctx,
+        if (isa<BuiltinProtocolConformance>(concrete))
+          return false;
+
         auto replacementTy = dc->mapTypeIntoContext(concrete->getType());
```
You won't need the `if` above once you fix the `getType()` invariant

kavon committed:
NCGenerics: break cycle with SuperclassTypeRequest

With NoncopyableGenerics, we get a cycle involving
`SuperclassTypeRequest` with this program:

  public struct RawMarkupHeader {}
  final class RawMarkup: ManagedBuffer<RawMarkupHeader, RawMarkup> { }

Because we generally don't support the following kind of relationship:

  class Base<T: P>: P {}
  class Derived: Base<Derived> {}

This commit works around the root-cause, which is that Derived's
synthesized conformance to Copyable gets superceded by the inherited one
from Base. Instead of recording conformances in the ConformanceLookup
table at all, create builtin conformances on the fly, since classes
cannot be conditionally Copyable or Escapable.

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test
